src/memray/_ipython.py

```diff
@@ -0,0 +1,169 @@
+                action="store_true",
+                help="Record allocations made by the Pymalloc allocator",
+                default=False,
+            )
+            alloc_type_group = parser.add_mutually_exclusive_group()
+            alloc_type_group.add_argument(
+                "--leaks",
+                help="Show memory leaks, instead of peak memory usage",
+                action="store_true",
+                dest="show_memory_leaks",
+                default=False,
+            )
+            alloc_type_group.add_argument(
+                "--temporary-allocation-threshold",
+                metavar="N",
+                help=dedent(
+                    """
+                    Report temporary allocations, as opposed to leaked allocations
+                    or high watermark allocations.  An allocation is considered
+                    temporary if at most N other allocations occur before it is
+                    deallocated.  With N=0, an allocation is temporary only if it
+                    is immediately deallocated before any other allocation occurs.
+                    """
+                ),
+                action="store",
+                dest="temporary_allocation_threshold",
+                type=int,
+                default=-1,
+            )
+            alloc_type_group.add_argument(
+                "--temporary-allocations",
+                help="Equivalent to --temporary-allocation-threshold=1",
+                action="store_const",
+                dest="temporary_allocation_threshold",
+                const=1,
+            )
+
+            parser.add_argument(
+                "--split-threads",
+                help="Do not merge allocations across threads",
+                action="store_true",
+                default=False,
+            )
+
+            return parser.parse_args(shlex.split(string))
+
+        @cell_magic  # type: ignore
+        def memray_flamegraph(self, line: str, cell: str) -> None:
+            """Memory profile the code in the cell."""
+            if self.shell is None:
+                raise UsageError("Cannot profile code when not in a shell")
+
+            options = self.parse_options(line)
+
+            results_dir = Path("memray-results")
+            if not results_dir.exists():
+                results_dir.mkdir()
+
+            tempdir = Path(tempfile.mkdtemp(dir=results_dir))
+            dump_file = Path(tempdir) / "memray.dump"
+            code = TEMPLATE.format(
+                tempdir=dump_file,
+                native_traces=options.native,
+                trace_python_allocators=options.trace_python_allocators,
+                follow_fork=options.follow_fork,
+                code=indent(cell, " " * 4),
+            )
+            self.shell.run_cell(code)
+
+            merge_threads = not options.split_threads
+
+            reporter = None
+
+            with FileReader(dump_file, report_progress=True) as reader:
```

Comment:
I am _extremely_ surprised to see that `report_progress=True` works in Jupyter notebooks. It's not perfect - it leaves extra blank lines between the cell's code and the output flamegraph where the progress bars were. But, a bit of extra wasted space is a small price to pay for the progress bar, so I think I'm happy to ship this despite the wasted space.